### DISCUSSION TITLE
improve json component DX

### DISCUSSION THREAD

**Discussion by mtt-artis** [sqlpage/SQLPage#609] [Ideas]
Hi 👋,

I like the json component but the forced use of `json(b)` functions add complexity not only because it requires extra knowledge but also because there are named differently across databases.

I'd like to propose something like this:

```sql
select 'json' as component;
select * from todos;
```
and 

```sql
select 'json' as component, true as first;
select * from todos where id = $id;
```

PS: love the project

**Comment by lovasoa** [comment]
Hi ! This would indeed be a very useful addition ! Would you be interested in implementing it ?

The json component is defined here:

https://github.com/lovasoa/SQLpage/blob/0a9390fd64e04b532f95c7ac2d36a3e8ca4a7664/src/render.rs#L189-L202

**Comment by lovasoa** [comment]
Hi @mtt-artis ! I implemented multiple improvements to the json component, including things you were suggesting above :smiley: 

Would you be interested in beta-testing it before we release v0.30 ? Development builds are available on docker as [`lovasoa/sqlpage:main`](https://registry.hub.docker.com/r/lovasoa/sqlpage/tags)

**Reply by ionv06** [reply]
Hello @lovasoa
The new json component is great, I'm waiting for a detailed example about 'sse'. I tried according to [https://sql.datapage.app/component.sql?component=json](https://sql.datapage.app/component.sql?component=json) but I have the error 'EventSource's response has a MIME type ("application/json") that is not "text/event-stream".'
L.E. Sorry, it's my mistake, I forgot about "'sse' as type'. It works! :)
What is the time interval at which the server retransmits the data to the client?

**Reply by lovasoa** [reply]
> What is the time interval at which the server retransmits the data to the client?

SQLPage will transmit the data as soon as it gets it from the database, so it depends on your query and your database.

**Reply by ionv06** [reply]
@lovasoa
I didn't ask the question correctly, I'm trying to rephrase. I have users.sql with the following content
```
select 'json' AS component, 'sse' AS type;
select username from users;
```
In index.sql 

`select 'shell' as component, 'index.js' as javascript`

And in index.js 
```
const eventSource = new EventSource("users.sql");
eventSource.onmessage = function (event) {
    const user = JSON.parse(event.data);
    console.log(user.username);
}
```
I access index.sql from the browser and open DevTools. In the console I see user names once, after a second - once again, then once more and so on; the instruction 'select username from users' is repeatedly sent to the mysql server. I'm missing something regarding the use of sse (in general, not just related to SQLPage).  Could I set a time interval at which to request the data from the server? Or an event related to the database (like a trigger) such as the insertion of a new record, and only then should the server send the data?

**Reply by lovasoa** [reply]
Thanks for the report, @ionv06 ! I updated the documentation to add:

```js
eventSource.onerror = function () {
    eventSource.close(); // do not reconnect after reading all the data
}
```

I also fixed a bug where the latest row was not sent properly.


Without the `eventSource.close()`, the browser will re-connect to the event source automatically whenever the server closes it (because it has sent all the query rows). It may be something you want, but in most cases, you want to fetch all the data just once, and you can close the event source on the browser as soon as the server closes it.

**Reply by ionv06** [reply]
@lovasoa
Fetching the data just once doesn't seem like what sse is supposed to do. 
I would like to know how I can send the data from the server periodically, at a certain time interval; how could I set an interval in users.sql?

**Reply by lovasoa** [reply]
the sse functionnality allows you to stream the results of a single query from the server to the client, a single time.

If you want to fetch the entire data at regular intervals, 
then you don't need server-sent events and can just poll the server regularly.
You can do so with:

### users.sql

```sql
select 'json' AS component;
select * from users;
```

### index.js

```js
async function fetch_users() {
  const req = await fetch('users.sql');
  const users = await req.json();
  console.log(users);
}

setInterval(fetch_users, 5*1000); // refetch users every 5 seconds forever
```

**Reply by ionv06** [reply]
Thanks for the example and the explanations. Still I don't get what would be the cases in which 'sse' would be useful. If it sends the result of a query to the client only once, isn't it the same as "select 'json' as component" (without 'sse' as type)?
I thought that 'sse' opens a communication channel from the server to the client through which the server sends notifications in a stream, without requests from the client (for example a task list generated from json content is updated on another device by another user, and the modification is visible to all clients in real time, without them sending requests)

**Reply by mtt-artis** [reply]
Hello.

SSE (Server-Sent Events) is a unidirectional communication protocol that allows the server to push updates to the client, unlike WebSockets, which support bidirectional communication (allowing the client to send messages to the server as well).

A potential use case for SSE in a Single Page Application (SPA) is streaming data, such as table rows, to be stored in the app's state or signals as they arrive. This enables the view to update incrementally, allowing users to see new rows in real-time without waiting for the entire payload to load before rendering.

**Reply by lovasoa** [reply]
> Still I don't get what would be the cases in which 'sse' would be useful. If it sends the result of a query to the client only once, isn't it the same as "select 'json' as component" (without 'sse' as type)?

SSE (Server-Sent Events) is useful when you want to send data from the server to the client progressively, as the data becomes available, instead of waiting for the entire query to finish.

Without SSE, if your query takes a long time to fetch all rows, the user will experience a delay (blank screen) until the entire dataset is retrieved and sent in one go (like your example with "select 'json' as component").

With SSE, as soon as the server has the first row, it can be pushed to the client immediately. The user can start seeing data even if the query is still running, improving perceived performance.

However, the current SSE implementation is not for "watching" ongoing changes to the data after the query is done (like subscribing to future changes). It’s just about delivering the results as soon as they are ready. That would be a great and useful feature, but it's not implemented yet.

**Reply by ToniJ123** [reply]
Hi, this sounds very interesting upgrade to SQLpage, is this possible to impelent in (near) future?

We have to limit at the moment very much amount of loaded data by filtering it because otherwise the page would be stuck for too long.

**Reply by lovasoa** [reply]
@ToniJ123 : Are you using the json component? I'm not sure how query watching in the json component relates to your problem of loading large amounts of data.

**Reply by ToniJ123** [reply]
> @ToniJ123 : Are you using the json component? I'm not sure how query watching in the json component relates to your problem of loading large amounts of data.

Yeah, sorry for misundestooding and confusion in this case, should have read better what we were talking about here, this isn't important section in our case, maybe some others benefit more from this 👍

**Comment by mtt-artis** [comment]
Hello @lovasoa 
Super hype about this feature.
Is the development build ready yet ?

**Reply by lovasoa** [reply]
Now they are!

**Comment by mtt-artis** [comment]
Love it !

Did some test on postgres with all the postgres type and it's working perfectly for select insert delete and update.
SSE is a nice feature for future use and more complex components.

Doing multiple query like the other components is a really cool trick.
```sql
select 'json' as component;
select id, name from cats;
select id, name from dogs;
select id, name from pokemon;
```

Having the db error in the json response is nice too.
```js
await fetch("/single.sql?id=no-one").then(r => r.json())
// {error: 'error returned from database: invalid input syntax for type integer: "no-one"'}
```

Quetions:
- In the documentation under the chapter "Send a single JSON object", should we warn users that both `'jsonlines' AS type` **AND** `LIMIT 1` are needed for correct behavior?

- For `jsonlines` that return no rows, an empty response body isn't valid JSON, leading to errors like below:
```js
await fetch("/single.sql?id=999").then(r => r.json())
// Uncaught SyntaxError: Unexpected end of JSON input
```
Should SQLPage provide a solution or pattern for this case? 
Like a fallbackStatus and fallbackBody option in the json component to handle cases where no rows are returned? 
I'm unsure about this, as I don't want to overcomplicate things and some libraries like already Redaxios handle this issue [Redaxios Source Code](https://github.com/developit/redaxios/blob/master/src/index.js#L213-L219). 

- Other edge cases I'm not certain sqlpage should handle as it's already the case for html pages
    - Should the status still return 200 even if there’s an error?
    - Should errors be stored in an array?
```js
await fetch("/all.sql?id=no-one").then(r => r.json())
// [{ "error": "error returned from database: invalid input syntax for type integer: \"no-one\"" }]
```

I'm really happy with the current implentation as it cover all my use cases (and more). **Thank you !**
